packages/dds/tree/src/test/feature-libraries/indexing/treeIndex.spec.ts

```diff
@@ -0,0 +1,255 @@
+				return anchorNodes.map((a) =>
+					getOrCreate(anchorIds, a, () => indexedAnchorNodeCount++),
+				);
+			},
+			(anchorNode: AnchorNode) => {
+				const simpleTree =
+					anchorNode.slots.get(proxySlot) ?? makeTreeNode(anchorNode, forest, root);
+				if (!isTreeValue(simpleTree)) {
+					return treeApi.status(simpleTree);
+				}
+			},
+		);
+
+		return {
+			index,
+			assertContents(...expected: [key: string, ...values: readonly TreeNode[]][]): void {
+				function assertSameElements(
+					actual: Iterable<unknown>,
+					expectedSet: Iterable<unknown>,
+				): void {
+					assert.deepEqual(new Set(actual), new Set(expectedSet));
+				}
+
+				const expectedEntries = expected.map(
+					([key, ...nodes]) =>
+						[
+							key,
+							nodes.map((f) => {
+								const flexNode: FlexTreeNode = getOrCreateInnerNode(f);
+								return getOrCreate(
+									anchorIds,
+									flexNode.anchorNode,
+									() => indexedAnchorNodeCount++,
+								);
+							}),
+						] as const,
+				);
+
+				// Check that the index reports the expected size
+				assert.equal(index.size, expectedEntries.length);
+
+				// Check that all expected entries are present
+				for (const [key, expectedNodes] of expectedEntries) {
+					assert.equal(index.has(key), true);
+					const nodes = index.get(key);
+					assert(nodes !== undefined);
+					assertSameElements(nodes, expectedNodes);
+				}
+
+				// Check that all iterators exactly match expected entries
+				assertSameElements(
+					index.keys(),
+					expectedEntries.map(([key]) => key),
+				);
+				assertSameElements(
+					index.values(),
+					expectedEntries.map(([_, value]) => value),
+				);
+				assertSameElements(index.entries(), expectedEntries);
+				assertSameElements(index, expectedEntries);
+
+				const set = new Map(expectedEntries);
+				index.forEach((value, key, i) => {
+					assert.equal(i, index);
+					assert.deepEqual(value, set.get(key));
+					set.delete(key);
+				});
+				assert.equal(set.size, 0);
+			},
+		};
+	}
+
+	it("can look up nodes in an initial tree", () => {
+		const { view, parent } = createView(new IndexableChild({ childKey: childId }));
+		const { assertContents } = createIndex(view);
+		const child = parent.child;
+		assert(child !== undefined);
+		assertContents([parentId, parent], [childId, child]);
+	});
+
+	it("can look up an inserted node", () => {
+		const { view, parent } = createView(); // create a parent with no child
+		const { assertContents } = createIndex(view);
+		parent.child = new IndexableChild({ childKey: childId });
+		const child = parent.child;
+		assert(child !== undefined);
+		assertContents([parentId, parent], [childId, child]);
+	});
+
+	// TODO: implement these series of tests
+	// they're not currently required because the API that will be made public makes this case impossible
+	// but the lower level API should still support and test this case
+	describe.skip("can re-index nodes in the spine of a replaced value", () => {
```

Comment:
is there an item filed for this?

Reply:
(or can we fix it now?)

Reply:
no item, I'm just doing it directly after in a separate pr because it feels like this one is getting a bit big